test/fft_real_to_complex.cpp

```diff
@@ -31,49 +32,91 @@ void print(const std::vector<T>& V)
 {
   using Real    = typename Backend::value_type;
   using Complex = boost::multiprecision::complex<Real>;
+  const Real tol = tolerance*std::numeric_limits<Real>::epsilon();
   
-  std::vector<Real> A{1.,2.,3.};
-  std::vector<Complex> B;
-    
-  Backend plan(A.size());
-  plan.real_to_complex(A.begin(),A.end(),std::back_inserter(B));
-  print(B);
+  boost::random::mt19937 rng;
+  boost::random::uniform_real_distribution<double> U(0.0,1.0);
   
-  A.push_back(4);
-  B.resize(A.size());
+  std::vector<Real> A(n);
+  std::vector<Complex> B(A.size());
   
-  plan.real_to_complex(A.begin(),A.end(),B.begin());
-  print(B);
-}
-
-template<typename Real>
-void test_bsl_fftw_gsl() {
-//test_r2c< bsl_rfft <Real> >();
-  test_r2c< fftw_rfft<Real> >();
-  test_r2c< gsl_rfft <Real> >();
-}
-
-template<typename Real>
-void test_bsl_fftw () {
-//test_r2c< bsl_rfft <Real> >();
-  test_r2c< fftw_rfft<Real> >();
-}
-
-template<typename Real>
-void test_bsl() {
-//test_r2c< bsl_rfft <Real> >();
+  for(unsigned int i=0;i<A.size();++i)
+  {
+    A[i] = U(rng);
+    B[i] = Complex{A[i],0.0};
+  }
+  
+  std::vector<Real> HC,iHC;
+  Backend rplan(A.size());
+  rplan.real_to_halfcomplex(A.begin(),A.end(),std::back_inserter(HC));
+  rplan.halfcomplex_to_real(HC.begin(),HC.end(),std::back_inserter(iHC));
+  //print(A);
+  //print(HC);
+  //print(iHC);
+  
+  std::vector<Complex> TB;
+  bsl_dft<Complex> cplan(B.size());
+  cplan.forward(B.begin(),B.end(),std::back_inserter(TB));
+  
+  // check if the inverse recovers the original array
+  {
+    Real diff{0.0};
+    for(unsigned int i=0;i<A.size();++i)
+    {
+      diff += std::abs(A[i]-iHC[i]);
+    }
+    diff /= A.size();
+    CHECK_MOLLIFIED_CLOSE(Real{0.0},diff,tol);
+  }
+  // check if the halfcomplex contains the non-redundant complex components
+  {
+    Real diff{0.0};
+    diff += std::abs(TB[0]-HC[0]);
+    for(int i=1,j=n-1;i<=j;++i,--j)
+    {
+      diff += std::abs(TB[i].real()-HC[i]);
+      
+      if(i<j)
+      diff += std::abs(TB[j].imag()-HC[j]);
+    }
+    diff /= n;
+    CHECK_MOLLIFIED_CLOSE(Real{0.0},diff,tol);
+  }
 }
 
 int main()
 {
-  test_bsl_fftw    <float      >();
-  test_bsl_fftw_gsl<double     >();
-  test_bsl_fftw    <long double>();
-
+  // corner cases
+  test_r2c<fftw_rfft<double>>(1);
+  test_r2c<gsl_rfft<double>>(1);
+  test_r2c<bsl_rfft<double>>(1);
+  
+  // primes 
+  for(auto n: std::vector<int>{2,3,5,7,11,13,17,19})
+  {
+    test_r2c<fftw_rfft<double>>(n,4);
+    test_r2c<gsl_rfft<double>>(n,16);
+    test_r2c<bsl_rfft<double>>(n,4);
+  }
+  
+  // powers of two
+  for(auto n: std::vector<int>{2,4,8,16,32,64,128})
+  {
+    test_r2c<fftw_rfft<double>>(n,4);
+    test_r2c<gsl_rfft<double>>(n,4);
+    test_r2c<bsl_rfft<double>>(n,4);
+  }
+  // composite
+  for(auto n: std::vector<int>{6,9,10,12,14,15,18,20,21,22,24,25,26,27,28,30})
+  {
+    test_r2c<fftw_rfft<double>>(n,8);
+    test_r2c<gsl_rfft<double>>(n,16);
+    test_r2c<bsl_rfft<double>>(n,4);
```

Comment:
Notice that the BSL precision is better than GSL, and as good as the FFTW.

Reply:
> need to multiply times 1/N

I am not totally certain on this, but I believe this is simply (or can be considered) a matter of convention. We might eventually want to scale internally on BSL in order to have the same output behavior on real-tom-half-complex. But, again, I think this can all be documented convention.

> BSL precision is better than GSL, and as good as the FFTW

Great job! That is a really difficult one to achieve since FFTW is also known for both its speed, robustness as well as precision. You have really created the foundation of a nice body of work/utilities for powerful FFT math Eduardo!

Nice!